js/modules/loki_rpc.js

```diff
@@ -31,8 +34,210 @@ const decryptResponse = async (response, address) => {
+    { name: 'HMAC', hash: 'SHA-256' },
+    key,
+    ephemeralSecret
+  );
+
+  const ciphertext = await window.libloki.crypto.EncryptGCM(
+    symmetricKey,
+    plaintext
+  );
+
+  return { ciphertext, symmetricKey, ephemeral_key: ephemeral.pubKey };
+};
+
+// Returns the actual ciphertext, symmetric key that will be used
+// for decryption, and an ephemeral_key to send to the next hop
+const encryptForDestination = async (node, payload) => {
+  // Do we still need "headers"?
+  const reqStr = JSON.stringify({ body: payload, headers: '' });
+
+  return encryptForNode(node, reqStr);
+};
+
+// `ctx` holds info used by `node` to relay further
+const encryptForRelay = async (node, nextNode, ctx) => {
+  const payload = ctx.ciphertext;
+
+  const reqJson = {
+    ciphertext: dcodeIO.ByteBuffer.wrap(payload).toString('base64'),
+    ephemeral_key: StringView.arrayBufferToHex(ctx.ephemeral_key),
+    destination: nextNode.pubkey_ed25519,
+  };
+
+  const reqStr = JSON.stringify(reqJson);
+
+  return encryptForNode(node, reqStr);
+};
+
+const BAD_PATH = 'bad_path';
+
+// May return false BAD_PATH, indicating that we should try a new
+const sendOnionRequest = async (reqIdx, nodePath, targetNode, plaintext) => {
+  log.info('Sending an onion request');
+
+  const ctx1 = await encryptForDestination(targetNode, plaintext);
+  const ctx2 = await encryptForRelay(nodePath[2], targetNode, ctx1);
+  const ctx3 = await encryptForRelay(nodePath[1], nodePath[2], ctx2);
+  const ctx4 = await encryptForRelay(nodePath[0], nodePath[1], ctx3);
+
+  const ciphertextBase64 = dcodeIO.ByteBuffer.wrap(ctx4.ciphertext).toString(
+    'base64'
+  );
+
+  const payload = {
+    ciphertext: ciphertextBase64,
+    ephemeral_key: StringView.arrayBufferToHex(ctx4.ephemeral_key),
+  };
+
+  const fetchOptions = {
+    method: 'POST',
+    body: JSON.stringify(payload),
+  };
+
+  const url = `https://${nodePath[0].ip}:${nodePath[0].port}/onion_req`;
+
+  // we only proxy to snodes...
+  process.env.NODE_TLS_REJECT_UNAUTHORIZED = '0';
+  const response = await nodeFetch(url, fetchOptions);
+  process.env.NODE_TLS_REJECT_UNAUTHORIZED = '1';
+
+  return processOnionResponse(reqIdx, response, ctx1.symmetricKey, true);
+};
+
+// Process a response as it arrives from `nodeFetch`, handling
+// http errors and attempting to decrypt the body with `sharedKey`
+const processOnionResponse = async (reqIdx, response, sharedKey, useAesGcm) => {
+  log.info(`(${reqIdx}) [path] processing onion response`);
+
+  // detect SNode is not ready (not in swarm; not done syncing)
+  if (response.status === 503) {
+    log.warn('Got 503: snode not ready');
+
+    return BAD_PATH;
+  }
+
+  if (response.status === 504) {
+    log.warn('Got 504: Gateway timeout');
+    return BAD_PATH;
+  }
+
+  if (response.status === 404) {
+    // Why would we get this error on testnet?
+    log.warn('Got 404: Gateway timeout');
+    return BAD_PATH;
+  }
+
+  if (response.status !== 200) {
```

Comment:
Shouldn't we check for a range of HTTP success values here?
```suggestion
  if (response.status < 200 || response.status > 299) {
```

Reply:
We only return 200 from SS for a successful request. If we get something else (even if it is 2xx), we'd probably want to know about it.

Reply:
👍 Sounds good to me then